oxide_core/quaternion.rs

```diff
@@ -0,0 +1,169 @@
+use crate::vector::vector3::*;
+use crate::vector::*;
+use std::ops;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct Quaternion {
+  pub x: f32,
+  pub y: f32,
+  pub z: f32,
+  pub w: f32,
+}
+
+pub trait VectorPart {
+  fn get_vector_part(&self) -> Vector3;
+}
+
+impl VectorPart for Quaternion {
+  fn get_vector_part(&self) -> Vector3 {
+    Vector3 {
+      x: self.x,
+      y: self.y,
+      z: self.z,
+    }
+  }
+}
+
+pub trait ScalarPart {
+  fn get_scalar_part(&self) -> f32;
+}
+
+impl ScalarPart for Quaternion {
+  fn get_scalar_part(&self) -> f32 {
+    self.w
+  }
+}
+
+pub fn id() -> Quaternion {
+  let one = 1.0f32;
+  let zero = 0.0f32;
+
+  Quaternion {
+    x: zero,
+    y: zero,
+    z: zero,
+    w: one,
+  }
+}
+
+impl Scale for Quaternion {
+  fn scale(&self, t: f32) -> Self {
+    Quaternion {
+      x: self.x * t,
+      y: self.y * t,
+      z: self.z * t,
+      w: self.w * t,
+    }
+  }
+}
+
+pub fn dot(a: Quaternion, b: Quaternion) -> f32 {
+  a.get_scalar_part() * b.get_scalar_part()
+    + a.get_vector_part().inner_product(&b.get_vector_part())
+}
+
+pub fn inverse(q: Quaternion) -> Quaternion {
+  q.scale(q.get_length().recip())
+}
+
+pub fn conjugate(a: Quaternion) -> Quaternion {
+  let vector_neg = a.get_vector_part().negate();
+  Quaternion {
+    x: vector_neg.x,
+    y: vector_neg.y,
+    z: vector_neg.z,
+    w: a.w,
+  }
+}
+
+impl Quaternion {
+  pub fn get_squared_length(&self) -> f32 {
+    self.w * self.w + self.get_vector_part().get_squared_length()
+  }
+  pub fn get_length(&self) -> f32 {
+    self.get_squared_length().sqrt()
+  }
+}
+
+pub fn rotate_vector(q: Quaternion, v: Vector3) -> Vector3 {
+  let vector_q = q.get_vector_part();
+  let t = vector_q.outer_product(&v).scale(2.0f32);
+  v + t.scale(q.w) + vector_q.outer_product(&t)
+}
```

Comment:
Can you replace this function with `rotate(angle_in_radius: f32)` like this?
```C++
template <typename T>
void Quaternion<T>::Rotate(T angleInRadians)
{
    Vector3<T> axis;
    T currentAngle;

    GetAxisAngle(&axis, &currentAngle);

    currentAngle += angleInRadians;

    Set(axis, currentAngle);
}
```

Reply:
Is it okay to change like this?
```suggestion
pub fn rotate(q: Quaternion, angle_in_radius: f32) -> Quaternion {
  let (axis, angle) = get_axis_angle(q);
  from_axis_angle(axis, angle + angle_in_radius)
}
```

Reply:
Yes, good.

Reply:
✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)